File: proxy-pattern/java/src/AudioPlayer.java
package src;

// AudioPlayer支持多种音频格式播放，相对简单，在同一个类实现
public class AudioPlayer {
  String fileType;
  String filePath;

  public AudioPlayer(String fileType, String filePath) {
    this.fileType = fileType;
    this.filePath = filePath;
  }

  public void play() {
    System.out.println("AudioPlayer path:" + filePath);
    if (fileType.equals("mp3")) {
      System.out.println("AudioPlayer::play() mp3");
    } else if (fileType.equals("mp4")) {
      System.out.println("AudioPlayer::play() mp4");
    }
  }
}

File: adapter-pattern/c/src/audio_player.c
#include "func.h"

/* 具体播放器实现类 */

void audio_player_play(AudioPlayer *player)
{
  printf("\r\n AudioPlayer::play() [file_type=%s, file_path=\"%s]",
         player->file_type, player->file_path);
  // audio playing
}

// AudioPlayer的构造函数
AudioPlayer *audio_player_constructor(char *file_type, char *file_path)
{
  MediaPlayer *player = (MediaPlayer *)malloc(sizeof(MediaPlayer));
  strncpy(player->file_type, file_type, 20);
  strncpy(player->file_path, file_path, 1000);
  player->set_file_type = &set_player_file_type;
  player->set_file_path = &set_player_file_path;
  AudioPlayer *audio_player = (AudioPlayer *)player;
  audio_player->play = &audio_player_play;
  return audio_player;
}